packages/java-edition/src/mcfunction/node/argument.ts

```diff
@@ -293,31 +293,155 @@ export interface FloatRangeNode extends core.AstNode {
 	value: [number | undefined, number | undefined]
 }
 
-export interface ItemNode extends core.AstNode {
-	type: 'mcfunction:item'
-	children: (core.ResourceLocationNode | nbt.NbtCompoundNode)[]
+export interface ItemStackNode extends core.AstNode {
+	type: 'mcfunction:item_stack'
+	children:
+		(core.ResourceLocationNode | ComponentListNode | nbt.NbtCompoundNode)[]
 	id: core.ResourceLocationNode
-	nbt?: nbt.NbtCompoundNode
+	components?: ComponentListNode // since 1.20.5
+	nbt?: nbt.NbtCompoundNode // until 1.20.5
 }
-export namespace ItemNode {
-	export function is(node: core.AstNode | undefined): node is ItemNode {
-		return (node as ItemNode | undefined)?.type === 'mcfunction:item'
+export namespace ItemStackNode {
+	export function is(node: core.AstNode | undefined): node is ItemStackNode {
+		return (node as ItemStackNode | undefined)?.type ===
+			'mcfunction:item_stack'
+	}
+
+	export function mock(
+		range: core.RangeLike,
+	): ItemStackNode {
+		const id = core.ResourceLocationNode.mock(range, { category: 'item' })
+		return {
+			type: 'mcfunction:item_stack',
+			range: core.Range.get(range),
+			children: [id],
+			id,
+		}
 	}
+}
 
-	export function mock(range: core.RangeLike, isPredicate: boolean): ItemNode {
+export interface ComponentListNode extends core.AstNode {
+	type: 'mcfunction:component_list'
+	children: core.PairNode<core.ResourceLocationNode, nbt.NbtNode>[]
+}
+
+export namespace ComponentListNode {
+	export function is(node: core.AstNode): node is ComponentListNode {
+		return (node as ComponentListNode).type === 'mcfunction:component_list'
+	}
+}
+
+export interface ItemPredicateNode extends core.AstNode {
+	type: 'mcfunction:item_predicate'
+	children: (
+		| core.ResourceLocationNode
+		| core.LiteralNode
+		| ComponentTestsNode
+		| nbt.NbtCompoundNode
+	)[]
+	id: core.ResourceLocationNode | core.LiteralNode
+	tests?: ComponentTestsNode // since 1.20.5
+	nbt?: nbt.NbtCompoundNode // until 1.20.5
+}
+export namespace ItemPredicateNode {
+	export function is(
+		node: core.AstNode | undefined,
+	): node is ItemPredicateNode {
+		return (node as ItemPredicateNode | undefined)?.type ===
+			'mcfunction:item_predicate'
+	}
+
+	export function mock(
+		range: core.RangeLike,
+	): ItemPredicateNode {
 		const id = core.ResourceLocationNode.mock(range, {
 			category: 'item',
-			allowTag: isPredicate,
+			allowTag: true,
 		})
 		return {
-			type: 'mcfunction:item',
+			type: 'mcfunction:item_predicate',
 			range: core.Range.get(range),
 			children: [id],
 			id,
 		}
 	}
 }
 
+export interface ComponentTestsNode extends core.AstNode {
+	type: 'mcfunction:component_tests'
+	children: ComponentTestsAnyOfNode[]
```

Comment:
I think there is a layer of children too much here. This `children` field will always have exactly zero or one child. Maybe the best way to solve this is by getting rid of this node, and using `ComponentTestsAnyOfNode` in `ItemPredicateNode`

Reply:
Originally `ComponentTestsAnyOfNode` was the one being returned, but IMO encapsulating it in a `ComponentTestsNode` is more readable and maybe less prone to breakage if the representation for the AST changes

Reply:
I disagree with that. We don't need unnecessary wrapper nodes. It also makes things like checkers and completers more complex for no reason.

Reply:
Hmm I'm not so sure. My main concern is future-proofing, since having things refer to `ComponentTestsNode` seems better than having to change all references from `ComponentTestsAnyOfNode` to `MaybeSomethingMoreComplicatedNode[]` if that ever happens. I doubt it's a big deal in either case so I'm OK with the change.